src/main/scala/nodes/learning/BlockWeightedLeastSquares.scala

```diff
@@ -270,4 +315,39 @@ object BlockWeightedLeastSquaresEstimator extends Logging {
     a
   }
 
+  def groupByClasses(
+      features: Seq[RDD[DenseVector[Double]]],
+      labels: RDD[DenseVector[Double]])
+    : (Seq[RDD[DenseVector[Double]]], RDD[DenseVector[Double]]) = {
+
+    val nClasses = labels.first.length.toInt
+    val hp = new HashPartitioner(nClasses)
+
+    val n = labels.partitions.length.toLong
+
+    // Associate a unique id with each item
+    // as Spark does not gaurantee two groupBy's will come out in the same order.
+    val shuffledLabels = labels.mapPartitionsWithIndex { case (k, iter) =>
+      iter.zipWithIndex.map { case (item, i) =>
+        val classIdx = item.toArray.indexOf(item.max)
+        (classIdx, (i * n + k, item))
+      }
+    }.partitionBy(hp).values.mapPartitions { part =>
+      part.toArray.sortBy(_._1).map(_._2).iterator
```

Comment:
Yeah this part sucks and is because of the way Spark does shuffles right now - So if you have keys from 2 blocks say block1, block2 coming in, they needn't come in with the same order each time. So for example in the first shuffle we might get block1 ++ block2 and in the next one block2 ++ block1